Tesselator minor coding fixes + Fix of a calculation error

rsm-gh opened the pull request:
+ Clang fixes
+ Performance fix by not duplicating the shape (&shape)
+ Fix when a shape has no geometry

## Summary by Sourcery

Addresses minor coding issues, enhances performance, and fixes a calculation error in the Tesselator module. It also adds exception handling for invalid deviation and mesh quality values.

Bug Fixes:
- Fixes a calculation error that occurs when a shape has no geometry, preventing potential crashes or incorrect tessellation results.
- Fixes memory leaks by ensuring that allocated memory for vertices, normals, and triangle indices is properly deallocated in the destructor and in edge computation.
- Fixes clang warnings.
- Fixes the shape duplication, improving performance by avoiding unnecessary copying of shape data during tessellation.

Enhancements:
- Improves code readability and maintainability by using nullptr instead of NULL for null pointers.
- Updates the ShapeTesselator constructor to take a reference to a TopoDS_Shape object, avoiding unnecessary copying of the shape.

Tests:
- Adds exception handling to ensure that deviation and mesh quality are greater than 0.

rsm-gh commented:
- I will provide a test file + a test script for the calculation error
- I'll verify if the performance improvement regarding `std::move` can be applied here, probably not, depending on the user intentions

tpaviot commented:
@rsm-sh thank you for this contribution. I wait for the test and your feedback before merging

tpaviot commented:
Another change I suggest: I didn't remember there were a trailing reference to the outdated pythonocc.org website. Can you change the url to https://github.com/tpaviot/pythonocc-core?

rsm-gh commented:
Hello @tpaviot ,

1) Yeah, I' ll change the url. 

2) Regarding the ` Bnd_Box` check , I think that maybe it is worth to revert the because  the Python debugger is quite good, and It may be more clean to have the exception.

I created a code to reproduce the error:
```python
from OCC.Extend.DataExchange import read_step_file_with_names_colors
from OCC.Display.WebGl.x3dom_renderer import X3DomRenderer

input_file = "/home/cadweb/Downloads/test.step"

data = read_step_file_with_names_colors(input_file)

shape = list(data.keys())[0]

renderer = X3DomRenderer(path="/home/cadweb/Desktop")
renderer.DisplayShape(shape)

```

And here' s the output with the "fix" :
```bash
cadweb@cadweb-dev:~/Desktop$ python3 test.py
Number of shapes at root : 1
Name : Chapa de Reforço v26
## x3dom webgl renderer - render axes/planes : True - axes/plane zoom factor : 1.0
There is no shape to tesselate
```
The file is generated without any error message...


and without the "fix":

```bash
cadweb@cadweb-dev:~/Desktop$ python3 test.py
Number of shapes at root : 1
Name : Chapa de Reforço v26
## x3dom webgl renderer - render axes/planes : True - axes/plane zoom factor : 1.0
Traceback (most recent call last):
  File "/home/cadweb/Desktop/test.py", line 13, in <module>
    renderer.DisplayShape(shape)
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/usr/lib/python3/dist-packages/OCC/Display/WebGl/x3dom_renderer.py", line 443, in DisplayShape
    x3d_exporter.compute()
    ~~~~~~~~~~~~~~~~~~~~^^
  File "/usr/lib/python3/dist-packages/OCC/Display/WebGl/x3dom_renderer.py", line 297, in compute
    shape_tesselator = ShapeTesselator(self._shape)
  File "/usr/lib/python3/dist-packages/OCC/Core/Tesselator.py", line 239, in __init__
    _Tesselator.ShapeTesselator_swiginit(self, _Tesselator.new_ShapeTesselator(aShape))
                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
RuntimeError: Standard_ConstructionError: Bnd_Box is void raised from method ShapeTesselator of class ShapeTesselator
```

what do you think?

3) I'll provide the  `std::move` tests in the following days


Regards

rsm-gh commented:
I have updated the code mostly to:

- `throw std::invalid_argument` on `ShapeTesselator::Tesselate` if `myDeviation` or `mesh_quality` are `<= 0`
- Added a method to get the deviation value: `ShapeTesselator::GetDeviation` (also in Python)
- Added an example of how to handle the error in the `X3DomRenderer` class:
```
if shape_tesselator.GetDeviation() <= 0:
            raise ValueError("The deviation is <= 0.")
```
and created a proper python `ValueError`

This allows to handle the problem gracefully with python:

```python
from OCC.Extend.DataExchange import read_step_file_with_names_colors
from OCC.Display.WebGl.x3dom_renderer import X3DomRenderer

input_file = "/home/cadweb/Downloads/test.step"

data = read_step_file_with_names_colors(input_file)

shape = list(data.keys())[0]

renderer = X3DomRenderer(path="/home/cadweb/Desktop")
renderer.DisplayShape(shape)
```
Will throw:
```
Number of shapes at root : 1
Name : 
## x3dom webgl renderer - render axes/planes : True - axes/plane zoom factor : 1.0
Traceback (most recent call last):
  File "/home/cadweb/Desktop/test.py", line 13, in <module>
    renderer.DisplayShape(shape)
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/usr/lib/python3/dist-packages/OCC/Display/WebGl/x3dom_renderer.py", line 447, in DisplayShape
    x3d_exporter.compute()
    ~~~~~~~~~~~~~~~~~~~~^^
  File "/usr/lib/python3/dist-packages/OCC/Display/WebGl/x3dom_renderer.py", line 300, in compute
    raise ValueError("The deviation is <= 0.")
ValueError: The deviation is <= 0.
```
But it can be managed as the following:

```
try:
	renderer.DisplayShape(shape)
except ValueError:
	pass
```	

Here there is a file for testing: [test.txt](https://github.com/user-attachments/files/19314677/test.txt) (to be renamed `.step` due to github limitations)


*The study regarding std::move stills on going.

rsm-gh commented:
Most of the comments of the AI bot do not actually concern my modifications, but the Swig comment is accurate, I directly modified the python interface file. Maybe @tpaviot you can suggest the change? I do not know swig

rsm-gh commented on src/Tesselator/ShapeTesselator.cpp:
```diff
@@ -271,18 +279,18 @@ void ShapeTesselator::ComputeEdges()
-        nbNodesInFace = aPoly->NbNodes();
+        nbNodesInFace = aPoly2->NbNodes();
         theEdge->number_of_coords = nbNodesInFace;
         theEdge->vertex_coord = new Standard_Real[nbNodesInFace * 3 * sizeof(Standard_Real)];
```
This was already on the original code, it is not coming from this modification.
